cmd/syncer/main.go

```diff
@@ -156,43 +170,93 @@ func startSyncers(ctx context.Context, GVRs []string, secretRef client.ObjectKey
-	if err != nil {
-		return nil, fmt.Errorf("unable to create rest config from control plane secret: %v", err)
-	}
 	upstreamClient, err := clusterSecret.DynamicClientsetFromSecret(controlPlaneSecret)
 	if err != nil {
 		return nil, fmt.Errorf("unable to create client from control plane rest config: %v", err)
 	}
 	informerFactory := dynamicinformer.NewDynamicSharedInformerFactory(upstreamClient, DEFAULT_RESYNC)
-	logger.Info("starting informer")
 	informerFactory.Start(ctx.Done())
-	logger.Info("waiting for cache sync")
 	informerFactory.WaitForCacheSync(ctx.Done())
-	logger.Info("cache sync complete")
 
+	//for spec: upstream is control-plane, downstream is data plane
 	specSyncConfig := syncer.Config{
-		UpstreamClientConfig: upstreamRestConfig,
-		GVRs:                 GVRs,
-		DownStreamClient:     mgr.GetClient(),
-		InformerFactory:      informerFactory,
-		ClusterID:            upstreamClusterConfig.Name,
-		NeverSyncedGVRs:      NEVER_SYNCED_GVRs,
+		GVRs:            GVRs,
+		InformerFactory: informerFactory,
+		ClusterID:       upstreamClusterConfig.Name,
+		NeverSyncedGVRs: NEVER_SYNCED_GVRs,
+		UpstreamNS:      controlPlaneNS,
+		DownstreamNS:    dataPlaneNamespace,
 	}
 
 	downstreamDynamicClient, err := dynamic.NewForConfig(mgr.GetConfig())
 	if err != nil {
 		return nil, fmt.Errorf("could not make dynamic downstream client from manager rest config: %v", err.Error())
 	}
-	SpecSyncer, err := spec.NewSpecSyncer(upstreamClusterConfig.Name, upstreamClusterConfig.Name, upstreamClient, downstreamDynamicClient)
+	SpecSyncer, err := spec.NewSpecSyncer(upstreamClusterConfig.Name, upstreamClusterConfig.Name, upstreamClient, downstreamDynamicClient, specSyncConfig)
 	if err != nil {
 		return nil, fmt.Errorf("could not create new spec syncer: %v", err.Error())
 	}
-	logger.Info("starting syncer", "name", upstreamClusterConfig.Name)
+	logger.Info("starting spec syncer", "name", upstreamClusterConfig.Name)
 
 	go SpecSyncer.Start(ctx)
 
-	syncRunnable := syncer.GetSyncerRunnable(ctx, specSyncConfig, syncer.InformerForGVR, SpecSyncer)
+	syncRunnable := syncer.GetSyncerRunnable(specSyncConfig, syncer.InformerForGVR, SpecSyncer)
+	return syncRunnable, nil
+
+}
+
+func startStatusSyncers(ctx context.Context, GVRs []string, secretRef client.ObjectKey, mgr manager.Manager, controlPlaneNS string) (*syncer.SyncRunnable, error) {
+	logger := log.FromContext(ctx)
+	controlPlaneSecret := &corev1.Secret{}
+	secretClient, err := client.New(mgr.GetConfig(), client.Options{})
+	if err != nil {
+		setupLog.Error(err, "unable to create sync secret client")
+		os.Exit(1)
+	}
+	err = secretClient.Get(ctx, secretRef, controlPlaneSecret, &client.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve control plane config secret '%v', error: %v", secretRef.String(), err)
+
+	}
+	downstreamClusterConfig, err := clusterSecret.ClusterConfigFromSecret(controlPlaneSecret)
```

Comment:
I am not following this, downstream is the cluster the sync agent is installed in?

Reply:
this ``` downstreamClient, err := clusterSecret.DynamicClientsetFromSecret(controlPlaneSecret) ``` reads strangely

Reply:
Downstream is where the sync is sync to. Upstream is where it is syncing from.

For Spec: upstream = control plane, downstream = data plane
for Status: upstream = data plane, downstream = control plane